desc_source.go

```diff
@@ -302,3 +294,75 @@ func addFilesToSet(allFiles []*descriptorpb.FileDescriptorProto, expanded map[st
 	}
 	return append(allFiles, fd.AsFileDescriptorProto())
 }
+
+// WriteProtoFiles will use the given descriptor source to resolve all the given
+// symbols and write proto files with their definitions to the given output directory.
+func WriteProtoFiles(outProtoDirPath string, descSource DescriptorSource, symbols ...string) error {
+	filenames, fds, err := getFileDescriptors(symbols, descSource)
+	if err != nil {
+		return err
+	}
+	// now expand that to include transitive dependencies in topologically sorted
+	// order (such that file always appears after its dependencies)
+	expandedFiles := make(map[string]struct{}, len(fds))
+	allFilesSlice := make([]*desc.FileDescriptor, 0, len(fds))
+	for _, filename := range filenames {
+		allFilesSlice = addFilesToFileDescriptorList(allFilesSlice, expandedFiles, fds[filename])
+	}
+	pr := protoprint.Printer{}
+	// now we can serialize to files
+	for _, fd := range allFilesSlice {
+		fdFQName := fd.GetFullyQualifiedName()
+		dirPath := filepath.Dir(fdFQName)
+		outFilepath := filepath.Join(outProtoDirPath, dirPath)
+		if err := os.MkdirAll(outFilepath, 0755); err != nil {
+			return fmt.Errorf("failed to create directory %q: %v", outFilepath, err)
+		}
+		fileName := filepath.Base(fdFQName)
+		filePath := filepath.Join(outFilepath, fileName)
+		f, err := os.Create(filePath)
+		if err != nil {
+			if f != nil {
+				_ = f.Close()
+			}
+			return fmt.Errorf("failed to create file %q: %v", filePath, err)
+		}
```

Comment:
immediately after this block, you should probably `defer f.Close()`

Reply:
you can't defer in a for loop.

See number 2: https://blog.learngoprogramming.com/gotchas-of-defer-in-go-1-8d070894cb01

Reply:
You have to scope it sometimes:
```go
for _, fd := range allFilesSlice {
	fdFQName := fd.GetFullyQualifiedName()
	dirPath := filepath.Dir(fdFQName)
	outFilepath := filepath.Join(outProtoDirPath, dirPath)
	if err := os.MkdirAll(outFilepath, 0755); err != nil {
		return fmt.Errorf("failed to create directory %q: %v", outFilepath, err)
	}
	fileName := filepath.Base(fdFQName)
	filePath := filepath.Join(outFilepath, fileName)
	err := func() error {
		f, err := os.Create(filePath)
		if err != nil {
			return fmt.Errorf("failed to create")
		}
		defer f.Close()
		if err := pr.PrintProtoFile(fd, f); err != nil {
			return fmt.Errorf("failed to write")
		}
		if err := f.Close(); err != nil {
			return fmt.Errorf("failed to close")
		}
		return nil
	}()
	if err != nil {
		return fmt.Errorf("file %q: %w", filePath, err)
	}
}
return nil
```

Reply:
^^ yeah this

Reply:
ready